### PR TITLE
[SPARK-36521][SQL] Disallow comparison between Interval and String

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Disallow comparison between Interval and String in the default type coercion rules.

### Why are the changes needed?

If a binary comparison contains interval type and string type, we can't decide which
interval type the string should be promoted as. There are many possible interval
types, such as year interval, month interval, day interval, hour interval, etc.

### Does this PR introduce _any_ user-facing change?

No, the new interval type is not released yet.

### How was this patch tested?

Existing UT

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 1]
```diff
@@ -862,6 +862,18 @@ object TypeCoercion extends TypeCoercionBase {
     case _ => None
   }
 
+  // Return whether a string literal can be promoted as the give data type in a binary comparison.
+  private def canPromoteAsInBinaryComparison(dt: DataType) = dt match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:866]: is it possible to share this code between ansi and non-ansi?

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 1 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:866]: Yes and no. Here is for binary comparison, while in ANSI mode it is for binary operators...It is hard to create a function name for this.

File: sql/core/src/test/resources/sql-tests/results/interval.sql.out [hunk 2]
```diff
@@ -2292,49 +2292,55 @@ cannot resolve '(INTERVAL '1' MONTH > INTERVAL '20' DAY)' due to data type misma
 -- !query
 SELECT INTERVAL '1' DAY < '1'
```
Comment (MaxGekk) [inline sql/core/src/test/resources/sql-tests/results/interval.sql.out:2293]: I wonder are there checks where the string at the left side?

Comment (gengliangwang) [reply sql/core/src/test/resources/sql-tests/results/interval.sql.out:2293]: Added new test cases.

**Review by cloud-fan — APPROVED** [review]

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master/3.2.
Thank you, @gengliangwang and @cloud-fan for review.

**Pull request closed without merging**